tiktorch/converters.py

```
@@ -33,6 +33,25 @@ class NamedImplicitOutputShape:
     halo: NamedShape
 
 
+@dataclasses.dataclass(frozen=True)
+class Tensor:
+    spec_id: str
+    data: xr.DataArray
+
+    def __hash__(self):
+        return hash(self.spec_id)
+
+    def __eq__(self, other):
+        if isinstance(other, Tensor):
+            return self.spec_id == other.spec_id
+        return False
```

Not sure I understand the motivation for this class. I see this is implemented for the sake if using it in a set, so that Data could be completely different, but if I say `t1 == t2` for two tensors that have the same `tensor_id` (the term `spec_id` is a bit misleading, as there are many different `ids` in the spec, e.g. in the future for the axes. Maybe just having a blank `id` field here would be nicer) then it's fine?

What I don't like about it is that `set` will fail silently if there are multiple tensors with the same id. So the one thing that could be going wrong (as opposed to how spec is parsed into prediction pipeline) is actually not checked. This should already be caught by the spec, and in my opinion, tiktorch should fail here, if a message is encountered with multiple tensors with the same id. I mean how do you decide which one to keep and send back. How can the client know what's going on?

---

I introduced this class mainly to do the client validation as mentioned here https://github.com/ilastik/tiktorch/pull/215#discussion_r1713846922. I thought that having just a tensor, without any metadata of what this tensor should represent from the model side isn't very convenient. For example a model could need multiple input tensors, each one of them with an id, so an incoming tensor need to be somehow associated with the model.

Actually this is the concept of sample, now I realized that! I could use the `Sample` object, that could keep a dict of the tensors, so with this way we can validate that the tensors are unique, and we skip the set implementation.

Definitely the check of `t1 == t2` is quite dangerous, that's why I have tried to add the `equal` method. I could argue that symbol `==` in python is yet another function, and dev should not make assumption on what means in objects.

Also I didn't get what you mean that `set` fails silently. You will get an error if the two tensors with same ids are added to the set, so the client understood that I messed up with ids. Then later on we don't know if the names are actually the ones expected, but still this is validated by the `ShapeValidator` implicitly when we check the shape and tensor id, it should exists in the names of the input tensors.